---
title: "Preprocessing for whole genome structural variant analyses"
author: "Robert Scharpf"
date: \today
output: BiocStyle::pdf_document
vignette: >
  %\VignetteIndexEntry{Preprocessing}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc} 
---

# Normalizing bin counts

We begin by loading several R packages required for this vignette.

```{r packages, results="hide"}
library(svfilters.hg19)
library(Rsamtools)
library(svbams)
library(trellis)
```

We begin by defining a set of bins over which the number of aligned reads will be counted.  The sv.filters<.build> packages contain 1kb bins that exclude regions with average mappability less than 0.75.
To make this toy example run faster, we only preprocess chromosome 8.

```{r define_bins}
data(bins1kb, package="svfilters.hg19")
bins <- keepSeqlevels(bins1kb, "chr8", pruning.mode="coarse")
```

Next, we extract a toy bam file from the svbams package.

```{r bamfile}
extdata <- system.file("extdata", package="svbams")
bamfile <- file.path(extdata, "CGOV44T.bam")
```

We use infrastructure provided in the Rsamtools package for counting single-end reads overlapping the bins.


```{r counts}
flags <- scanBamFlag(isDuplicate=FALSE,
                     isPaired=TRUE,
                     isUnmappedQuery=FALSE,
                     hasUnmappedMate=FALSE,
                     isProperPair=TRUE)
bviews <- BamViews(bamPaths=bamfile, bamRanges=bins)
bins$cnt <- binnedCounts(bviews)
bins <- bins[ bins$cnt > 0 ]
```

Next, we log-transform the raw counts and correct for GC content by loess. As a random subset of the genome is selected to model GC biases, we set a seed in the following code chunk for reproducibility.

```{r normalize}
bins$std_cnt <- binNormalize(bins)
set.seed(123)
bins$gc_adj <- binGCCorrect(bins)
```

At this stage, it is a good idea to plot the data.  When many samples are available, differences between groups of samples (batch effects) can be visualized and addressed at this stage.

```{r plot}
library(ggplot2)
tibble(chr = "chr8", start=start(bins),
       end=end(bins), log2_ratio=bins$gc_adj) %>%
  filter(!is.na(log2_ratio)) %>%
  ggplot(.)  +
  geom_point(aes(start, log2_ratio), size = 1) +
  xlab("Coordinate") +
  ylab("log2 normalized counts") +
  coord_cartesian(ylim=c(-4, 1)) +
  geom_hline(yintercept=0)
```

If one or more normal controls are available, the bin-counts can be further corrected by the bin-wise medians from normals.  This is particularly helpful for capture-based sequencing where bins are often of variable width.
