---
title: "Overview of svcnvs package"
author: "Robert Scharpf"
date: \today
output: BiocStyle::pdf_document
vignette: >
  %\VignetteIndexEntry{Overview of svcnvs package}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc} 
---

# Introduction

# Workflow

This vignette requires the *svpreprocess* package for loading a views for the
bin-level preprocessed coverage estimates.

```{r packages}
library(svpreprocess)
library(svcnvs)
```

```{r views}
data(pview, package="svpreprocess")
```

Update the file path in the views object to file containing the preprocessed coverage.

```{r views_filepath}
ddir <- system.file("extdata", package="svpreprocess", mustWork=TRUE)
cov.file <- file.path(ddir, "preprocessed_coverage.rds")
pview@bamPaths <- cov.file
```

TODO: show how deletions are called from the views object above.

Here, we are loading results from segmentation of the preprocessed coverage as
well as an object of called deletions from the *sv_deletions* function for the
ovarian cancer cell line CGOV44T.

```{r deletion}
data(segments, package="svcnvs")
data(deletion, package="svcnvs")
```

# Plotting deletions

We will use the *ggplot2* and *gridExtra* packages for plotting the deletions.

```{r plot_libraries}
library(ggplot2)
library(gridExtra)
```

In the following code chunk, we extract the genomic coordinates for the deletion
stored in the *deletion* object. To view the deletion in the context of the
surrounding region, we create a second *GRanges* object that includes 200kb of
the flanking genome on each side of the deletion.

```{r region_of_interest}
library(SummarizedExperiment)
##
## region of interest (roi)
##
roi <- variant(deletion)
roi <- keepSeqlevels(roi, as.character(seqnames(roi)))
expand <- 200e3
roi2 <- GRanges(seqnames(roi), IRanges(start(roi)-expand,
                                       end(roi) + expand))
```

Next, we subset the views object to contain only the genomic bins in the region
of interest defined above. In addition, we create a data.frame containing all
the segments for this particular chromosome and sample and a data.frame
containing the preprocessed coverage.

```{r dataframes, eval=FALSE}
pview2 <- pview[queryHits(findOverlaps(rowRanges(pview), roi2)), ]
segs <- keepSeqlevels(segments, seqlevels(roi), pruning.mode="coarse")
segs.df <- as(segs, "data.frame")
df <- data.frame(logr=assays(pview2)[, 1],
                 start=start(rowRanges(pview2)))
```

We restrict the y-axis limits to a suitable range for visualizing the log
ratios, thresholding log ratios that are extreme. We highlight the region
identified by the segmentation in the ggplot graphic and use dashed lines to
indicate the more precise deletion boundaries identified from the rearranged
read pairs that were used in the *sv_deletion* function. 

```{r plot_logratios}
ylim <- c(-9, 2)
df$logr <- threshold(df$logr, ylim)
brks <- pretty(df$start, n=8)
region <- subsetByOverlaps(segs, roi)
region <- region[region$seg.mean < -1]
region <- as.data.frame(region)
xlim <- c(start(roi2), end(roi2))

A <- ggplot(df, aes(start, logr)) +
  geom_point(size=1, color="gray50") +
  scale_x_continuous(expand=c(0,0), breaks=brks, labels=brks/1e6)+
  scale_y_continuous(expand=c(0,0)) +
  geom_segment(data=segs.df,
               aes(x=start, xend=end, y=seg.mean, yend=seg.mean),
               size=1) +
  coord_cartesian(xlim=xlim, ylim=ylim) +
  ylab(expression(log[2]~ratio)) +
  geom_rect(data=region,
            aes(xmin=start, xmax=end, ymin=-Inf, ymax=+Inf),
            fill="steelblue", color="transparent", alpha=0.3,
            inherit.aes=FALSE) +
  theme(axis.text=element_text(size=10),
        axis.text.x=element_blank()) + xlab("") +
  annotate("text", x=xlim[1] + 15e3, y=-8, label="chr15", size=3)
A1 <- ggplotGrob(A)
```

We would like to visualize the log ratio






