---
title: "Overview of svfusions package"
author: "Robert Scharpf"
date: \today
output: BiocStyle::pdf_document
vignette: >
  %\VignetteIndexEntry{Overview of svfusions package}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc} 
---

# Introduction

Nearly all sequence junctions formed by a rearrangement occur in non-coding regions of the genome. With whole genome sequencing, we can identify the location of the noncoding sequence junctions. When these junctions occur within a gene (usually intronic), we can evaluate whether the new transcript produced by the rearrangement is in-frame. This vignette provides an overview, extending the rearrangement analyses in the `svrearrange` package.

Our analysis to identify in-frame fusions begin with a `RearrangementList` for ovarian cell line `CGOV11T\_1`  created by and available from the `svrearrange` package.

# Analysis pipeline

```{r RearrangementList}
suppressPackageStartupMessages(library(svrearrange))
##load_all("svrearrange")
##suppressPackageStartupMessages(library(svfusions))
extdata <- system.file("extdata", package="svrearrange")
rfile <- file.path(extdata, "CGOV11T_1.bam.rds")
rlist <- readRDS(rfile)
rlist
```

The above `rlist` contains 68 rearrangements. We begin by selecting only the rearrangements in which both ends of a sequence junction are within 5kb of a known transcript.


```{r rlist_subset}
build <- "hg19"
near.coding <- seqJunctionNearTx(rlist, build)
rlist <- rlist[ near.coding ]
length(rlist)
```

As discussed in the vignette for the `svrearrange` package, each candidate rearrangement has two possible 5-prime to 3-prime orientations that can be inferred by the orientation of the rearranged read pairs (RRPs) and split reads (SRs).  We use the function `fiveTo3List` to order each rearrangement in the `rlist` object by these two orientations.  In doing so, the length of the `rlist` object doubles.

```{r fiveToThree}
rlist2 <- fiveTo3List(rlist, build)
length(rlist2)
```

Having established the 5-prime to 3-prime orientations in the `rlist2` object, we make the definition of the sequence junctions more precise.

```{r seq_junctions}
jxns <- seqJunctions_Rlist(rlist2)
jxns
```

In addition, we require the 3-prime genomic region of the junction to lie strictly within a transcript, while the 5-prime genomic region forming the junction can occur either in the promoter or within the 5-prime transcript.  Operationally, we define the promoter as the genomic region 5kb upstream of the transcription start site of the 5-prime gene. We refer to the remaining junctions as `coding_jxns`.  As the coding junctions are named by the rearrangement ids, we can use these names to  subset the `rlist2` object, excluding those rearrangements that do not meet the above criteria.

```{r coding_jxns}
coding_jxns <- codingJunctions(jxns, build)
coding_jxns
rlist2 <- rlist2[ names(coding_jxns) ]
```

Next, we use the `fuseCDS_Rlist` function to extract for each rearrangement the full CDS of the fused sequence in the somatic genome (`fusions`), the partial CDS of the 5-prime (`tum.5p`) and 3-prime (`tum.3p`) transcripts that are involved in the fusion, and the full CDS of the 5-prime (`ref.5p`) and 3-prime transcripts in the reference genome. To speed up computation in the remaining portions of this vignette, we restrict the `RearrangementList` object to a rearrangement involving the fusion of the tyrosine kinase IKAROS to a known driver ERBB4.  Note, we use a single bracket `[` when subsetting the `RearrangementList` so that the resulting object is still an instance of `RearrangementList` (albeit a length-one list).

```{r fuseCDS}
ikaros.erbb4 <- "9136-9181"
coding_jxns <- coding_jxns[ikaros.erbb4]
cds.list <- fuseCDS_Rlist(rlist2[ikaros.erbb4], coding_jxns)
cds.list
```

For each fusion, we translate the CDS in each of the three possible reading frames and partition the amino acid sequence of the fusion protein into the sequences derived from the 5-prime and 3-prime transcripts.  In particular, by partioning the amino acid sequence of the fusion into its 5-prime and 3-prime parts, we can compare the 5-prime partition to the amino acid sequence of the 5-prime reference protein and the 3-prime partition to the amino acid sequence of the 3-prime reference protein. The function `translateCDS` translates the CDS in the 3 possible reading frames. 

```{r translate}
proteins <- translateCDS(cds.list)
proteins
```

The amino acid sequence obtained by translating the CDS involved in the 5-prime and 3-prime transcripts are represented as `AAStringSet` objects.  Since a single gene can have multiple transcripts, we translate every possible combination of 5-prime and 3-prime transcripts in each of the 3 frames.  Here, we show the amino acid sequences for the two possible combinations of transcripts involving IKAROS and ERBB4:

```{r aminoacid_sequence}
proteins$fusion[[1]]
```

We can also list the amino acid sequences derived from the 5-prime gene and the 3-prime gene:

```{r aminoacid_sequence_5prime}
proteins$tum5p[[1]]
proteins$tum3p[[1]]
```

The function `partitionAASequence` reorganizes the list of amino acid sequences by frame to facilitate downstream computation.

```{r partitionAASequence}
partition.fusion <- partitionAASequence(proteins)
partition.fusion
```

For a fusion to be in-frame, we require that the amino acid sequence derived from the 5- and 3-prime transcripts to be subsequences of the full amino acid sequence of the reference genome that was translated in the same frame.  In addition, we require that there be no premature stop codons.  The following code accomplishes these two tasks, with the `ref.frames` object below containing the full amino acid sequence of the transcripts associated with IKAROS and ERBB4.  

```{r inframe_and_nostops}
nostop.list <- noPrematureStop(partition.fusion)
nostop.list
ref.frames <- organizeReferenceByFrame(proteins)
inframe.list <- inFrameList(fusion.frames=partition.fusion,
                            ref.frames=ref.frames)
inframe.list
```

Each element of the `nostop.list` and `inframe.list` lists are an object of class `LogicalList`, each evaluated from a different reading frame.  The `LogicalList` objects are named by the rearrangement identifier.  The helper function `inFrameNoStop` combines these to lists to create a single list that is `TRUE` for rearrangements that are in-frame and have no premature stop codons.

```{r combine_LogicalLists}
inframe.nostop <- inFrameNoStop(nostop.list, inframe.list)
inframe.nostop
```

Finally, `validFusions` returns a list object containing only the fusions that are in-frame and have no premature stops.  For each such fusion, the amino acid sequence, CDS of the fusion, partial CDS of the 5-prime and 3-prime transcripts, and the full CDS of the reference transcripts are available.

```{r valid_fusions}
valid.fusions <- validFusions(partition.fusion,
                              cds.list,
                              inframe.nostop,
                              ref.frames)
valid.fusions
```

The above data can be summarized in tabular format using the `fusionTable2` function:

```{r fusionTable2}
tab <- fusionTable2(valid.fusions)
head(tab)
```

Note, this table provides both the genomic coordinates of the fusion (junction.5prime, junction.3prime), but also the portion of the amino acid sequence that was retained by the fusion (aa.5prime.start, aa.5prime.end, aa.3prime.start, aa.3prime.end).  In the following code chunk, we construct amino acid ranges for the genes involved in the fusion and the protein domains for these genes in the reference genome.

```{r protein_domains}
  tumor_aa_ranges <- aa_granges(tab)
  extdata2 <- system.file("extdata", package="svfusions")
  up <- readRDS(file.path(extdata2, "uniprot.rds"))
  up2 <- uniprotFeatures(up, tab, strwrap.width=20)
  domain_aa_ranges <- GRanges(up2$hugo, IRanges(up2$start, up2$end),
                              chromosome=up2$seqnames,
                              description=up2$description,
                              short.desc=up2$short.desc,
                              aa_len=up2$aa_len)
  domains <- subsetByOverlaps(domain_aa_ranges, tumor_aa_ranges, type="within")
  domains
```
